support/default_output_validator.cpp

```diff
@@ -2,199 +2,379 @@
-			}
-		} else {
-			while(s >> w) words.push_back(w);
+// each token is either EOF, a single whitespace charachter or a string without any whitespace character
+struct token_view {
+	std::string_view token;
+
+	token_view() {}
+	token_view(std::string_view token) : token{token} {}
+
+	bool is_eof() const {
+		return token.empty();
+	}
+
+	std::optional<char> is_space() const {
+		if (token.size() != 1) return {};
+		if (not util::is_space(token[0])) return {};
+		return token[0];
+	}
+
+	std::optional<long double> is_float() const {
+		if (not util::is_float(token)) return {};
+		try {
+			//std::from_chars for floats is not widely supported ):
+			std::size_t pos = 0;
+			std::string tmp(token);
+			long double res = std::stold(tmp, &pos);
+			assert(pos == token.size());
+			assert(std::isfinite(res));
+			return res;
+		} catch(const std::out_of_range& /**/) {
+			// i dont know how to handle this... return +-inf?
+			return {};
+		} catch(...) {}
+		// this should not happen. Except for the range check our parsing is stricter than std::stold
+		assert(false);
+		return {};
+	}
+
+	std::string formatted(std::size_t lim = 200) const {
+		assert(lim >= TEXT_ELLIPSIS.size());
+		if (is_eof()) return "EOF";
+		if (token == " ") return "\" \"";
+		if (token == "\f") return "\"\\f\"";
+		if (token == "\n") return "\"\\n\"";
+		if (token == "\r") return "\"\\r\"";
+		if (token == "\t") return "\"\\t\"";
+		if (token == "\v") return "\"\\v\"";
+		std::string res(token);
+		if (res.size() > lim) {
+			res.resize(lim - TEXT_ELLIPSIS.size());
+			res += TEXT_ELLIPSIS;
 		}
-		return words;
-	};
-	const auto ans_words = words(ans);
-	const auto out_words = words(out);
-
-	if(ans_words == out_words) {
-		assert(not space_change_sensitive);
-		return {true, "white space"};
-	}
-
-	if(floatabs == 0 and floatrel == 0) {
-		return {false, quick_diff(out, ans)};
-	}
-
-	if(out_words.size() != ans_words.size()) {
-		return {false, quick_diff(out, ans)};
-	}
-
-	long double max_abs_err = 0;
-	long double max_rel_err = 0;
-	for(int i = 0; i < out_words.size(); ++i) {
-		const auto& w1 = ans_words[i];
-		const auto& w2 = out_words[i];
-		if(w1 != w2) {
-			size_t p1 = 0, p2 = 0;
-			// If the answer term doesn't parse as a float, don't try the output term.
-			// In this case, we always need equality of w1 and w2.
-			long double v1, v2;
-			try {
-				v1 = stold(w1, &p1);
-			} catch(exception& e) {
-				return {false, quick_diff(w2, w1)};
-			}
-			if(p1 < w1.size()) return {false, quick_diff(w2, w1)};
+		return res;
+	}
 
-			// If the output term doesn't parse as a float -> WA.
-			try {
-				v2 = stold(w2, &p2);
-			} catch(exception& e) {
-				return {false, quick_diff(out, ans)};
-			}
-			if(p2 < w2.size()) return {false, quick_diff(w2, w1)};
+	bool equal(token_view o) const {
+		return token == o.token;
+	}
 
-			// OK if w1 and w2 represent the same (possibly nan/inf) value.
-			if(v1 == v2) continue;
+	bool insensitive_equal(token_view o) const {
```

Comment:
rename to `case_insensitive_equal` (as opposed to 'float_insensitive_equal' or so)